crates/shared/src/price_estimation/http.rs

```diff
@@ -419,6 +429,99 @@ mod tests {
+                        }),
+                        fee: None,
+                        exec_plan: None,
+                    }
+                },
+                amms: hashmap! {
+                    0 => UpdatedAmmModel {
+                        execution: vec![ExecutedAmmModel {
+                            sell_token: H160::from_low_u64_be(0),
+                            buy_token: H160::from_low_u64_be(1),
+                            exec_sell_amount: 100.into(),
+                            exec_buy_amount: 100.into(),
+                            exec_plan: None
+                        },ExecutedAmmModel {
+                            sell_token: H160::from_low_u64_be(1),
+                            buy_token: H160::from_low_u64_be(0),
+                            exec_sell_amount: 100.into(),
+                            exec_buy_amount: 100.into(),
+                            exec_plan: None
+                        }],
+                        cost: Some(TokenAmount {
+                            amount: 200_000.into(),
+                            token: native_token
+                        }
+                        ),
+                    }
+                },
+                interaction_data: vec![InteractionData {
+                    target: H160::zero(),
+                    value: U256::zero(),
+                    call_data: vec![],
+                    inputs: vec![],
+                    outputs: vec![],
+                    exec_plan: None,
+                    cost: Some(TokenAmount {
+                        amount: 300_000.into(),
+                        token: native_token,
+                    }),
+                }],
+                ..Default::default()
+            })
+        });
+
+        let mut token_info_fetching = MockTokenInfoFetching::new();
+        token_info_fetching
+            .expect_get_token_infos()
+            .returning(move |_| HashMap::new());
+
+        let gas_price_estimating = Arc::new(FakeGasPriceEstimator::new(GasPrice1559 {
+            max_fee_per_gas: 1.0,
+            max_priority_fee_per_gas: 1.0,
+            ..Default::default()
+        }));
+
+        let estimator = HttpPriceEstimator::new(
+            Arc::new(api),
+            Arc::new(FakePoolFetcher(vec![])),
+            None,
+            None,
+            Arc::new(token_info_fetching),
+            gas_price_estimating,
+            native_token,
+            Arc::new(BaseTokens::new(native_token, &[])),
+            "test".into(),
+            RateLimiter::test(),
+        );
+
+        let query = Query {
+            from: None,
+            sell_token: H160::from_low_u64_be(0),
+            buy_token: H160::from_low_u64_be(1),
+            in_amount: 100.into(),
+            kind: OrderKind::Sell,
+        };
+        let result = estimator.estimate(&query).await.unwrap();
+
+        // 94391 base cost + 100k order cost + 200k AMM cost (x2) + 300k interaction cost
```

Comment:
The 2x is a bit counter-intuitive. If the AMM is touched twice in the solution, the top level cost is multiplied (it's assumed the cost is denominated per execution). I don't think this is true in practice for concentrated liquidity or even weighted pools (where depending on in/out token the cost can differ).

Not that it matter that much in practice (and I didn't want to change the semantics), but a subtlety to be a ware of (cc @harisang)